File: src/cro/crot.f90
! THIS VERSION: GALAHAD 3.3 - 28/05/2021 AT 13:30 GMT.
   PROGRAM GALAHAD_CRO_TEST   ! ** to be improved!
   USE GALAHAD_CRO_double                    ! double precision version
   IMPLICIT NONE
   INTEGER, PARAMETER :: wp = KIND( 1.0D+0 ) ! set precision
   REAL ( KIND = wp ), PARAMETER :: infinity = 10.0_wp ** 20
   TYPE ( CRO_data_type ) :: data
   TYPE ( CRO_control_type ) :: control        
   TYPE ( CRO_inform_type ) :: inform
   INTEGER :: i
   INTEGER, PARAMETER :: n = 11, m = 3, m_equal = 1, a_ne = 30, h_ne = 21
   INTEGER, DIMENSION( h_ne ) :: H_col
   INTEGER, DIMENSION( n + 1 ) :: H_ptr
   REAL ( KIND = wp ), DIMENSION( h_ne ) :: H_val
   INTEGER, DIMENSION( a_ne ) :: A_col
   INTEGER, DIMENSION( m + 1 ) :: A_ptr
   REAL ( KIND = wp ), DIMENSION( a_ne ) :: A_val
   REAL ( KIND = wp ), DIMENSION( n ) :: G, X_l, X_u, X, Z
   REAL ( KIND = wp ), DIMENSION( m ) :: C_l, C_u, C, Y
   INTEGER, DIMENSION( m ) :: C_stat
   INTEGER, DIMENSION( n ) :: X_stat
! start problem data
   H_val = (/ 1.0D+0, 5.0D-1, 1.0D+0, 5.0D-1, 1.0D+0, 5.0D-1, 1.0D+0, 5.0D-1,  &
              1.0D+0, 5.0D-1, 1.0D+0, 5.0D-1, 1.0D+0, 5.0D-1, 1.0D+0, 5.0D-1,  &
              1.0D+0, 5.0D-1, 1.0D+0, 5.0D-1, 1.0D+0 /) ! H values
   H_col = (/ 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 10,    &
              11 /)                              ! H columns
   H_ptr = (/ 1, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22 /) ! pointers to H col
   A_val  = (/ 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, &
               1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, &
               1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, &
               1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0 /) ! A values
   A_col = (/ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 3, 4, 5, 6, 7, 8, 9, 10, 11,  &
              2, 3, 4, 5, 6, 7, 8, 9, 10, 11 /) ! A columns
   A_ptr = (/ 1, 12, 21, 31 /)                  ! pointers to A columns
   G   = (/ 5.0D-1, -5.0D-1, -1.0D+0, -1.0D+0, -1.0D+0,  -1.0D+0, -1.0D+0,     &
           -1.0D+0, -1.0D+0, -1.0D+0, -5.0D-1 /) ! objective gradient
   C_l = (/  1.0D+1, 9.0D+0, - infinity /)       ! constraint lower bound
   C_u = (/  1.0D+1, infinity, 1.0D+1 /)         ! constraint upper bound
   X_l = (/ 0.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0,                    &
            1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0 /) ! variable lower bound
   X_u  = (/ infinity, infinity, infinity, infinity, infinity, infinity,       &
             infinity, infinity, infinity, infinity, infinity /) ! upper bound
   C = (/ 1.0D+1, 9.0D+0, 1.0D+1 /)              ! optimal constraint value
   X = (/ 0.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0, 1.0D+0,      &
          1.0D+0, 1.0D+0, 1.0D+0 /)              ! optimal variables
   Y = (/  -1.0D+0, 1.5D+0, -2.0D+0 /)           ! optimal Lagrange multipliers
   Z = (/ 2.0D+0, 4.0D+0, 2.5D+0, 2.5D+0, 2.5D+0, 2.5D+0,                      &
          2.5D+0, 2.5D+0, 2.5D+0, 2.5D+0, 2.5D+0 /) ! optimal dual variables
   C_stat = (/ -1, -1, 1 /)                         ! constraint status
   X_stat = (/ -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1 /) ! variable status
! problem data complete
   CALL CRO_initialize( data, control, inform ) ! Initialize control parameters
   CALL CRO_crossover( n, m, m_equal, H_val, H_col, H_ptr, A_val, A_col,       &
                       A_ptr, G, C_l, C_u, X_l, X_u, C, X, Y, Z, C_stat,       &
                       X_stat , data, control, inform )  ! crossover
   IF ( inform%status == 0 ) THEN                   ! successful return
     WRITE( 6, "( '      x_l          x          x_u          z     stat', /,  &
   &               ( 4ES12.4, I5 ) )" )                                        &
       ( X_l( i ), X( i ), X_u( i ), Z( i ), X_stat( i ), i = 1, n )
     WRITE( 6, "( '      c_l          c          c_u          y     stat', /,  &
   &               ( 4ES12.4, I5 ) )" )                                        &
       ( C_l( i ), C( i ), C_u( i ), Y( i ), C_stat( i ), i = 1, m )
     WRITE( 6, "( ' CRO_solve exit status = ', I0 ) " ) inform%status
   ELSE                                            ! error returns
     WRITE( 6, "( ' CRO_solve exit status = ', I0 ) " ) inform%status
   END IF
   CALL CRO_terminate( data, control, inform )     ! delete internal workspace
   END PROGRAM GALAHAD_CRO_TEST

File: src/cro/C/crot.c
/* crot.c */
/* Full test for the CRO C interface using C sparse matrix indexing */

#include <stdio.h>
#include <math.h>
#include "galahad_precision.h"
#include "galahad_cfunctions.h"
#include "galahad_cro.h"

int main(void) {

    // Derived types
    void *data;
    struct cro_control_type control;
    struct cro_inform_type inform;


    // Set problem dimensions
    int n = 11; // dimension
    int m = 3; // number of general constraints
    int m_equal = 1; // number of equality constraints

    //  describe the objective function

    int H_ne = 21;
    real_wp_ H_val[] = {1.0,0.5,1.0,0.5,1.0,0.5,1.0,0.5,1.0,0.5,
                        1.0,0.5,1.0,0.5,1.0,0.5,1.0,0.5,1.0,0.5,1.0};
    int H_col[] = {0,0,1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9,10};
    int H_ptr[] = {0,1,3,5,7,9,11,13,15,17,19,21};
    real_wp_ g[] = {0.5,-0.5,-1.0,-1.0,-1.0, -1.0,-1.0,-1.0,-1.0,-1.0,-0.5};

    //  describe constraints

    int A_ne = 30;
    real_wp_ A_val[] = {1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,
                        1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,
                        1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0};
    int A_col[] = {0,1,2,3,4,5,6,7,8,9,10,2,3,4,5,6,7,8,9,10,
                   1,2,3,4,5,6,7,8,9,10};
    int A_ptr[] = {0,11,20,30};
    real_wp_ c_l[] = {10.0,9.0,-INFINITY};
    real_wp_ c_u[] = {10.0,INFINITY,10.0};
    real_wp_ x_l[] = {0.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0};
    real_wp_ x_u[]  = {INFINITY,INFINITY,INFINITY,INFINITY,INFINITY,INFINITY,
                       INFINITY,INFINITY,INFINITY,INFINITY,INFINITY};

    // provide optimal variables, Lagrange multipliers and dual variables
    real_wp_ x[] = {0.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0, 1.0,1.0,1.0};
    real_wp_ c[] = {10.0,9.0,10.0};
    real_wp_ y[] = { -1.0,1.5,-2.0};
    real_wp_ z[] = {2.0,4.0,2.5,2.5,2.5,2.5,2.5,2.5,2.5,2.5,2.5};
    // provide interior-point constraint and variable status
    int c_stat[] = {-1,-1,1};
    int x_stat[] = {-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1};

    // Set output storage

    char st;
    int status;

    printf(" C sparse matrix indexing\n\n");

    // Initialize CRO
    cro_initialize( &data, &control, &status );

    // Set user-defined control options
    control.f_indexing = false; // C sparse matrix indexing

    // crossover the solution
    cro_crossover_solution( &data, &control, &inform, 
                            n, m, m_equal, 
                            H_ne, H_val, H_col, H_ptr, 
                            A_ne, A_val, A_col, A_ptr, 
                            g, c_l, c_u, x_l, x_u, x, c, y, z, 
                            x_stat, c_stat );

    printf(" CRO_crossover exit status = %1i\n", inform.status);

    // Delete internal workspace
    cro_terminate( &data, &control, &inform );

}
